### PR TITLE
Documentation for cfg dominator class

### PR DESCRIPTION
**Pull request by danpoe**

### CODE REVIEW

File: src/analyses/cfg_dominators.h [hunk 1]
```diff
@@ -62,8 +68,8 @@ class dominatorst
 
 public:
   /// Create an empty set
-  /// Note: Only unreachable nodes should be assigned
-  /// empty sets after the algorithm completes
+  /// Note: Only unreachable nodes or the root node should be assigned empty
```
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/cfg_dominators.h]: This isn't true, the root node dominates itself (it doesn't have an immediate dominator though).

File: src/analyses/cfg_dominators.h [hunk 2]
```diff
@@ -180,26 +186,22 @@ class dominatorst
...
-    // failure, so we use this as an ugly workaround until that code is fixed.
-
-    // NOLINTNEXTLINE
+    // FIXME This works around a bug in other parts of the code in particular,
+    // dependence_graph.cpp, where iterators to different lists than those that
+    // are stored in this set are passed to find. The Debug libstdc++ will
+    // (correctly!) run into an assertion failure using std::find. std::less for
+    // some reason doesn't trigger this assertion failure, so we use this as an
+    // ugly workaround until that code is fixed.
+	// NOLINTNEXTLINE
```
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/cfg_dominators.h]: Can you restore the empty line and the indentation here?

Comment (hannes-steffenhagen-diffblue) [inline src/analyses/cfg_dominators.h]: Can you restore the empty line and the indentation here? The empty line was there to separate the explanation from the linter hint.

File: src/analyses/cfg_dominators.h [hunk 3]
```diff
@@ -219,6 +221,7 @@ template <typename T, typename node_indext>
 const node_indext
   dominatorst<T, node_indext>::npos = std::numeric_limits<node_indext>::max();
 
+/// Dominators for each node
```
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/cfg_dominators.h]: Maybe something like "Compute dominators for each instruction in a GOTO program"?

File: src/analyses/cfg_dominators.h [hunk 4]
```diff
@@ -265,12 +270,34 @@ class cfg_dominators_templatet
...
     std::vector<std::set<node_indext>> bucket;
 
+    // Used by link() and eval(), which are used to create and query
+    // an auxiliary data structure which is a forest that is contained
+    // in the DFS spanning tree.
+	std::vector<node_indext> ancestor;
+    std::vector<node_indext> label;
+    std::vector<node_indext> size;
+	std::vector<node_indext> child;
+
```
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/cfg_dominators.h]: :+1:

File: src/analyses/cfg_dominators.h [hunk 5]
```diff
@@ -414,38 +459,51 @@ class cfg_dominators_templatet
...
+/// Lengauer and Tarjan. Node indices are 1-based as in the paper, with the
+/// first element (with index 0) of each data structure simply left empty.
 template <class P, class T, bool post_dom>
 void cfg_dominators_templatet<P, T, post_dom>::fixedpointt::fixedpoint(
   P &program)
 {
-  // Dominator Tree according to Lengauer and Tarjan
-  // "A fast algorithm for finding dominators in a flow graph"
-  // This is ununderstandable without reading the paper!
-  // assumption: Vertex indices >= 0 and < cfg.size()
```
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/cfg_dominators.h:424]: In hindsight, "ununderstandable" was probably not the right choice of words :slightly_smiling_face:

File: src/analyses/cfg_dominators.h [hunk 6]
```diff
@@ -414,38 +459,51 @@ class cfg_dominators_templatet
...
+  // Assumption: Vertex indices > 0 and <= cfg.size()
```
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/cfg_dominators.h]: In cfg vertex indices are 0 based and `< cfg.size()`; Can you clarify if you're talking about the algorithm (for which this is true) or cfg (for which this is not true) here?

File: src/analyses/cfg_dominators.h [hunk 7]
```diff
@@ -500,23 +565,40 @@ void cfg_dominators_templatet<P, T, post_dom>::initialise(P &program)
...
 inline void dominators_pretty_print_node(
   const goto_programt::targett &target,
   std::ostream &out)
 {
   out << target->code.pretty();
 }
 
+/// Pretty-print a single node.
+/// \param node: node to print
+/// \param out: output stream
```
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/cfg_dominators.h:594]: Not sure if there's much meaning to parameter documentation if it doesn't carry more information than the type. Then again, it doesn't hurt either.

**Review by hannes-steffenhagen-diffblue — COMMENTED** [review]
Generally happy with these changes, can you address my comments? After that, we can put this in. I think Michael will be happy enough with this.

File: src/analyses/cfg_dominators.h (hunk 1 above)
Comment (danpoe) [reply src/analyses/cfg_dominators.h]: Ah right.

File: src/analyses/cfg_dominators.h (hunk 6 above)
Comment (danpoe) [reply src/analyses/cfg_dominators.h]: Yes, I'll add a clarification.

**Pull request merged**